### PR TITLE
Add gcp and idtoken authenticators

### PR DESCRIPTION
**Pull request by asetty**

### CODE REVIEW

**Review by makofok — APPROVED** [review]

File: main.go [hunk 1]
```diff
@@ -148,6 +152,11 @@ func main() {
 		oauth2Config:            oauth2Config,
 	}
 
+	gcpAuthenticator, err := newGcpAuthenticator(c.GCPHeader)
+	if err != nil {
+		log.Fatalf("error initializing gcp authenticator")
```
Comment (makofok) [inline main.go:157]: It would be useful to also emit the actual error
```
log.Fatalf("error initializing gcp authenticator: %v", err)
```

**Review by makofok — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (makofok) [reply main.go:157]: Actually, do we want it to totally fail if it cannot load the gcp authenticator? Or just log and move on...

```
time="2021-08-31T09:05:06Z" level=fatal msg="error initializing gcp authenticator: error creating validator: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information."
```

Comment (makofok) [reply main.go:157]: Actually failing on load isn't great either, since then you will have problems in server.go:103 at runtime...

Comment (makofok) [reply main.go:157]: maybe we can have something like this in main.go
```
	authenticators := []Authenticator{sessionAuthenticator, idTokenAuthenticator}
	if gcpAuthenticator != nil {
		authenticators = append(authenticators, gcpAuthenticator)
	}

...

		caBundle:                caBundle,
		authenticators:          authenticators,
		authorizers:             []Authorizer{groupsAuthorizer},
```

File: go.mod [hunk 2]
```diff
@@ -14,14 +14,13 @@ require (
...
+	github.com/stretchr/testify v1.6.1
 	github.com/tevino/abool v0.0.0-20170917061928-9b9efcf221b5
 	github.com/yosssi/boltstore v1.0.1-0.20150916121936-36632d491655
 	golang.org/x/exp v0.0.0-20201008143054-e3b2a7f2fdc7 // indirect
-	golang.org/x/net v0.0.0-20201021035429-f5854403a974 // indirect
-	golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
-	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	golang.org/x/oauth2 v0.0.0-20210427180440-81ed05c6b58c
 	gonum.org/v1/gonum v0.8.1
+	google.golang.org/api v0.46.0
```
Comment (makofok) [inline go.mod:23]: I hit https://github.com/googleapis/google-api-go-client/pull/1198 pretty hard while trying to get this branch working locally.

Need to update to v.0.56.0

```
go get google.golang.org/api
...
go get: upgraded google.golang.org/api v0.46.0 => v0.56.0

go mod tidy
```

**Comment by makofok** [comment]
So I eventually got the googleSAAuthenticator working. See previous notes. For reference I used this method (https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oidc) to create the idtoken.

I noticed that the current implementation seems to accept _any_ service account:
```
time="2021-09-01T05:11:13Z" level=info msg="UserInfo: &{Name:dummy-client@sw-infra-auth-sandbox.iam.gserviceaccount.com Groups:[]}" host="localhost:8080" ip="192.168.16.1,192.168.16.4" method=GET path=/qotm/
```

**Comment by kellyma2** [comment]
Closing this PR - we're going to use client credentials flow for M2M authentication and we'll pull in the idtoken authenticator as part of our updated master branch

**Pull request closed without merging**